### PR TITLE
add SetExtraData to CodecParameters to set extradata 

### PR DESCRIPTION
**Pull request by Cacsjep**
For publishing an rtmp stream extradata needs to be set (PPS, SPS h264).

### CODE REVIEW

File: codec_parameters.go [hunk 1]
```diff
@@ -141,6 +145,27 @@ func (cp *CodecParameters) SetSampleAspectRatio(r Rational) {
...
+		return nil
+	}
+
+	if cp.c.extradata != nil {
+		C.av_freep(unsafe.Pointer(&cp.c.extradata))
+		cp.c.extradata_size = 0
+	}
+
+	extradataSize := len(extraData)
+	cp.c.extradata = (*C.uint8_t)(C.av_mallocz(C.size_t(extradataSize + C.AV_INPUT_BUFFER_PADDING_SIZE)))
```
Comment (asticode) [inline codec_parameters.go:159]: Could you shorten with
```go
if cp.c.extradata = ...; cp.c.extradata == nil {
   ...
}
```

File: codec_parameters.go [hunk 2]
```diff
@@ -141,6 +145,27 @@ func (cp *CodecParameters) SetSampleAspectRatio(r Rational) {
...
+	if cp.c.extradata == nil {
+		return fmt.Errorf("failed to allocate extradata")
```
Comment (asticode) [inline codec_parameters.go:161]: Could you update the error message with `astiav: allocation is nil` instead ?

File: codec_parameters.go [hunk 3]
```diff
@@ -141,6 +145,27 @@ func (cp *CodecParameters) SetSampleAspectRatio(r Rational) {
...
+	}
+
+	C.memcpy(unsafe.Pointer(cp.c.extradata), unsafe.Pointer(&extraData[0]), C.size_t(extradataSize))
+	cp.c.extradata_size = C.int(extradataSize)
+	return nil
+}
+
```
Comment (asticode) [inline codec_parameters.go:168]: Could you add a `func (cp *CodecParameters) ExtraData() ([]byte, error)` method as well (using `bytesFromC` like [here](https://github.com/asticode/go-astiav/blob/master/packet.go#L89))?

File: codec_parameters_test.go [hunk 4]
```diff
@@ -63,6 +63,9 @@ func TestCodecParameters(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, 2, cp5.Channels())
 
+	err = cp5.SetExtraData([]byte{0, 0, 0, 1})
```
Comment (asticode) [inline codec_parameters_test.go:66]: Could you use `cp6` (at the bottom of the test function) instead to test the new methods and simplify with `require.NoError(t, cp6.SetExtraData([]byte{0, 0, 0, 1}))`

File: codec_parameters_test.go [hunk 5]
```diff
@@ -63,6 +63,9 @@ func TestCodecParameters(t *testing.T) {
...
+	require.NoError(t, err)
```
Comment (asticode) [inline codec_parameters_test.go:67]: Could you test `cp6.Extradata()` as well?

**Review by asticode — COMMENTED** [review]
Thanks for the PR!

Out of curiosity, would you mind sharing the extradata you needed to publish your rtmp stream? 🤔

**Comment by Cacsjep** [comment]
done =)

**Comment by Cacsjep** [comment]
When pushing a h264 rtmp stream you need to set the [H264 SPS/PPS](https://www.cardinalpeak.com/blog/the-h-264-sequence-parameter-set) as extra data, its like meta data, so that decoder knows how it can decode it correctly.  =)

**Comment by asticode** [comment]
Thanks for the PR! ❤️ 

Let me know whether you need a tag 👍

**Comment by asticode** [comment]
> When pushing a h264 rtmp stream you need to set the [H264 SPS/PPS](https://www.cardinalpeak.com/blog/the-h-264-sequence-parameter-set) as extra data, its like meta data, so that decoder knows how it can decode it correctly. In my case a I got a h264 bytestream from a IP-Camera the first frame is an IDR Frame, this includes the SPS,PPS.

Ha you mean that you're using the `Packet.FromData()` method and therefore need to manually set the stream's codec parameters extradata whereas when using a FormatContext, it gets filled automatically?

**Comment by Cacsjep** [comment]
> Ha you mean that you're using the `Packet.FromData()` method and therefore need to manually set the stream's codec parameters extradata whereas when using a FormatContext, it gets filled automatically?

correct =)

**Comment by Cacsjep** [comment]
> Let me know whether you need a tag 👍
thanks no tag needed =)

**Pull request merged**